### PR TITLE
Start Self Heal daemon if the storage type is Replica3

### LINKED ISSUE #40: Start 'shd' container if 'Replica3' is choosen type.
Currently in `server/server.py` there is no self-heal daemon started. Would be great to have it running.

### PR DESCRIPTION
**Pull request by aravindavk**
Start self heal daemon in each server pod if the storage type
is specified as Replica3.

Signed-off-by: Aravinda VK <mail@aravindavk.in>
Fixes: #40

### CODE REVIEW

File: operator/main.py [hunk 1]
```diff
@@ -141,7 +141,7 @@ def update_config_map(core_v1_client, obj):
         data["bricks"].append({
             "brick_path": "/bricks/%s/data/brick" % volname,
             "node": get_brick_hostname(volname, storage["node"]),
-            "node_id": str(uuid.uuid1()),
+            "node_id": storage["node_id"],
```
Comment (amarts) [inline operator/main.py:142]: this is changed now? (or should be a conflict somewhere here..)

File: server/glusterfsd.py [hunk 2]
```diff
@@ -152,13 +152,11 @@ def start():
...
             "-p", "/var/run/gluster/glusterfsd-%s.pid" % brick_path_name,
-            # TODO: Change socket file name
-            "-S", "/var/run/gluster/b99981c29a4c396c.socket",
+            "-S", "/var/run/gluster/brick.socket",
             "--brick-name", brick_path,
             "-l", "-",  # Log to stderr
-            # TODO: Change Node ID
             "--xlator-option",
-            "*-posix.glusterd-uuid=6958dddc-1842-4ee0-92df-b6a060dfba5e",
+            "*-posix.glusterd-uuid=%s" % os.environ["NODEID"],
```
Comment (amarts) [inline server/glusterfsd.py:159]: NOTE TO SELF: need to understand this better (for myself). option has 'glusterd' in it, and we should understand why and what it does.

**Comment by aravindavk** [comment]
Testing and CI changes yet to be done

File: server/shd.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    """
+    volname = os.environ["VOLUME"]
+
+    volfile_path = os.path.join(VOLFILES_DIR, "glustershd.vol")
+    generate_shd_volfile(volfile_path, volname)
+
+    # Send Analytics Tracker
+    # The information from this analytics is available for
+    # developers to understand and build project in a better way
+    send_analytics_tracker("server")
```
Comment (amarts) [inline server/shd.py]: considering this is another container in server pod, we can skip this IMO

File: server/shd.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+
+def generate_shd_volfile(volfile_path, volname):
+    """Generate Client Volfile for Glusterfs Volume"""
+    info_file_path = os.path.join(VOLINFO_DIR, "%s.info" % volname)
+    data = {}
+    with open(info_file_path) as info_file:
+        data = json.load(info_file)
+
+    template_file_path = os.path.join(TEMPLATES_DIR, "Replica3.shd.vol.j2")
```
Comment (amarts) [inline server/shd.py:27]: with this feature, should we disable client self-heal altogether? or is it ok to keep it for couple of releases?

File: server/shd.py (hunk 3 above)
Comment (aravindavk) [reply server/shd.py]: ack

File: server/glusterfsd.py (hunk 2 above)
Comment (aravindavk) [reply server/glusterfsd.py:159]: May be to identify local node ID

File: server/shd.py (hunk 4 above)
Comment (aravindavk) [reply server/shd.py:27]: I am not sure, we can keep it if no perf penalty.

**Comment by aravindavk** [comment]
Need to think about the support for Self heal status and trigger self heal without having CLIs

File: server/shd.py (hunk 4 above)
Comment (amarts) [reply server/shd.py:27]: There is some perf penalty for sure, I am not aware at present, how much.

**Comment by amarts** [comment]
> Need to think about the support for Self heal status and trigger self heal without having CLIs

How about a tool which deploys a container, and does `find` on mount point on PV (or on the whole volume itself). Like our sample-app-test container, which just does read/write, and lists mount details ?

**Comment by amarts** [comment]
Some thoughts:

Or a future sidecar container, which would be listening on a port, which can have a button from prometheus/grafana dashboard to trigger all this?

Or similar to current quotad.py which keeps watching for a file, and depending on the content, does some actions? like self-heal, like rebalance (future), or like a md5sum data check, etc etc?

**Comment by amarts** [comment]
```

 File "/usr/local/lib64/python3.7/site-packages/jinja2/parser.py", line 59, in fail
    raise exc(msg, lineno, self.name, self.filename)
jinja2.exceptions.TemplateSyntaxError: Unexpected end of template. Jinja was looking for the following tags: 'elif' or 'else' or 'endif'. The innermost block that needs to be closed is 'if'.

```

Also operator crashlooped 4 times, and there are 'stale mount' warning.

**Comment by aravindavk** [comment]
Still failing on CI. Checking

**Comment by aravindavk** [comment]
Tests are passing now. Had an issue with shd volfile and server YAML files.

Tests are not covering the actual heal functionality, we may have to add those tests. I will see how to simulate the pod failures and testing self-heal later.

**Review by amarts — APPROVED** [review]

**Pull request merged**